File: day08.py
#!/usr/bin/env pipenv run python
"""Solutions to day 8 of Advent of Code"""

from get_input import get_input, line_parser

class Node:
    """Node in a tree with children and values
    """
    def __init__(self, meta, children):
        self.meta = meta
        self.children = children

    def __repr__(self):
        return f"Node({self.meta}, {self.children})"

    def value(self):
        """Value of the node equal to total if there are no children otherwise
        equal othe sum of the value of children referenced by cardinal index in
        meta
        """
        if self.children == tuple():
            return sum(self.meta)
        total = 0
        for meta in self.meta:
            if 0 < meta <= len(self.children):
                total += self.children[meta-1].value()
        return total

    def total(self):
        """Sum of meta plus the total of all child nodes"""
        return sum(self.meta) + sum(child.total() for child in self.children)

    @classmethod
    def build(cls, records):
        """Returns a tree and the number of items read from records"""
        children = []
        i = 2
        for _ in range(records[0]):
            j, child = cls.build(records[i:])
            i += j
            children.append(child)
        return (i + records[1]), cls(tuple(records[i:i+records[1]]), tuple(children))

def part1(records):
    """Solution to part 1"""
    _, root = Node.build(records)
    return root.total()

def part2(records):
    """Solution to part 2"""
    _, root = Node.build(records)
    return root.value()

if __name__ == '__main__':
    LINES = line_parser(get_input(day=8, year=2018), seperator=' ')
    print("Part 1: {}".format(part1(LINES)))
    print("Part 2: {}".format(part2(LINES)))

File: day08.rs
extern crate common;

use common::get_input;

type Input = Node;

#[derive(Debug)]
struct Node {
    meta: Vec<u32>,
    children: Vec<Node>,
}

impl Node {
    fn new(values: &mut Iterator<Item=u32>) -> Node {
        let n_children = values.next().unwrap();
        let n_meta = values.next().unwrap();
        let mut root = Node { meta: Vec::new(), children: Vec::new() };
        for _ in 0..n_children {
            root.children.push(Node::new(values));
        }
        for _ in 0..n_meta {
            root.meta.push(values.next().unwrap());
        }
        return root;
    }

    fn sum_meta(&self) -> u32 {
        let child_sum: u32 = self.children.iter().map(|v| v.sum_meta()).sum();
        let meta_sum: u32 = self.meta.iter().sum();
        return meta_sum + child_sum;
    }

    fn value(&self) -> u32 {
        if self.children.is_empty() {
            return self.sum_meta()
        }
        let mut total = 0;
        for &n in &self.meta {
            println!("{:?}, {:?}", n, self.children.get(1 + n as usize));
            if let Some(child) = self.children.get(1 + n as usize) {
                total += child.value();
            }
        }
        return total
    }
}

fn part1(input: &Input) -> u32 {
    input.sum_meta()
}

fn part2(input: &Input) -> u32 {
    input.value()
}

fn parse(input: String) -> Input {
    let mut values = input.trim().split(" ")
        .map(|v| v.parse().unwrap());
    return Node::new(&mut values)
}

fn main() {
    let input = parse(get_input(08, 2018));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
